### PR TITLE
feat(users): Add email domain based restriction for dashboard entry APIs

### LINKED ISSUE #6939: feat(users): Add email domain based restriction for dashboard entry APIs
Currently Zurich uses okta and they use a specific `auth_id` to see the okta login method.

If any user doesn't use `auth_id`, then the user will see the default auth methods that Hyperswitch has and they will be use them to login, which we should restrict for zurich users.

### PR DESCRIPTION
**Pull request by ThisIsMani**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [x] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

There will be a mapping of email domain and auth methods.
This PR will restrict users to enter into the dashboard based on the email domain.

### Additional Changes

- [x] This PR modifies the API contract
- [x] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

Closes #6939.

## How did you test it?

1. Create user auth method with email domain
    ```bash
    curl --location 'http://localhost:8080/user/auth' \
    --header 'Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjoiMjAxYTJlZWQtNGE0ZC00MjZlLTg1N2ItMjNhM2ZjZTk5NTMzIiwibWVyY2hhbnRfaWQiOiJtZXJjaGFudF8xNjk3NjIzMjI0Iiwicm9sZV9pZCI6Im1lcmNoYW50X2FkbWluIiwiZXhwIjoxNjk3Nzk2MDI3LCJvcmdfaWQiOiJvcmdfVzJ2NmtPZ2lMSWx3YnFoaVh4VUIifQ.HnkHHWboPt82-VlvE5GVCjTJ1sA82-dMbUxeYNl-mxk' \
    --header 'Content-Type: application/json' \
    --header 'api-key: ••••••' \
    --data '{
        "owner_id": "test",
        "owner_type": "organization",
        "auth_method": {
            "auth_type": "password",
            "private_config": {
                "base_url": "url",
                "client_id": "client_id",
                "client_secret": "client_secret"
            },
            "public_config": {
                "name": "okta"
            }
        },
        "allow_signup": true,
        "email_domain": "gmail.com" 
    }'
    ```
    Response will be 200 OK.

2. Update user auth method
    1. Auth Method
        ```bash
        curl --location --request PUT 'http://localhost:8080/user/auth' \
        --header 'Content-Type: application/json' \
        --header 'api-key: test_admin' \
        --header 'Cookie: Cookie_1=value' \
        --data '{
            "auth_method": {
                "id": "799596e6-e6ba-405c-b5cf-70fd510409dc",
                "auth_method": {
                    "auth_type": "open_id_connect",
                    "private_config": {
                        "base_url": "base url",
                        "client_id": "client id",
                        "client_secret": "client secret"
                    },
                    "public_config": {
                        "name": "okta"
                    }
                }
            }
        }
        '
        ```
        Response will be 200 OK.
    2. Email Domain
         ```bash
          curl --location --request PUT 'http://localhost:8080/user/auth' \
          --header 'Content-Type: application/json' \
          --header 'api-key: test_admin' \
          --header 'Cookie: Cookie_1=value' \
          --data '{
              "email_domain": {
                  "owner_id": "test",
                  "email_domain": "juspay.com"
              }
          }'
          ```
          Response will be 200 OK.

All the following APIs will restrict the users based on the email domain and the corresponding auth method
1. Signin - Password
2. Signup - Password
3. Connect Account - MagicLink
4. Forgot Password - Password
5. Send Verification Email - MagicLink
6. SSO Signin - OpenIdConnect
7. Terminate Auth Select - Selected auth method

If the auth method list for email domain is empty of consists of the above auth method, the request will be continued, or else the api will be stopped.

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/router/src/types/domain/user/user_authentication_method.rs [hunk 1]
```diff
@@ -14,4 +14,5 @@ pub static DEFAULT_USER_AUTH_METHOD: Lazy<UserAuthenticationMethod> =
         allow_signup: true,
         created_at: common_utils::date_time::now(),
         last_modified_at: common_utils::date_time::now(),
+        email_domain: String::from("default"),
```
Comment (apoorvdixit88) [inline crates/router/src/types/domain/user/user_authentication_method.rs:17]: won't it be same as auth_id?

File: crates/router/src/core/user.rs [hunk 2]
```diff
@@ -2585,14 +2708,20 @@ pub async fn terminate_auth_select(
...
+    let auth_methods = state
+        .store
+        .list_user_authentication_methods_for_email_domain(user_email.extract_domain()?)
+        .await
+        .change_context(UserErrors::InternalServerError)?;
+
+    let user_authentication_method = match (req.id, auth_methods.is_empty()) {
+        (Some(id), _) => auth_methods
+            .into_iter()
+            .find(|auth_method| auth_method.auth_id == id)
```
Comment (apoorvdixit88) [inline crates/router/src/core/user.rs:2721]:
```suggestion
            .find(|auth_method| auth_method.id == id)
```

**Review by apoorvdixit88 — APPROVED** [review]

**Review by inventvenkat — APPROVED** [review]

**Review by racnan — APPROVED** [review]
This can be later changed to have email domain as a nullable column instead.
We shouldn't force users to always add a domain based restriction.

File: crates/router/src/core/user.rs [hunk 3]
```diff
@@ -2374,44 +2434,100 @@ pub async fn update_user_authentication_method(
...
+                    let user_error = match error.current_context() {
+                        StorageError::ValueNotFound(_) => {
+                            UserErrors::InvalidAuthMethodOperationWithMessage(
+                                "Auth method not found".to_string(),
+                            )
+                        }
+                        StorageError::DuplicateValue { .. } => {
+                            UserErrors::UserAuthMethodAlreadyExists
+                        }
+                        _ => UserErrors::InternalServerError,
```
Comment (SanchithHegde) [inline crates/router/src/core/user.rs:2471]: Can we have some context message included wherever we're raising internal server errors (here and elsewhere in the file)?

**Pull request merged**